elastic/elastic-agent
# Create installation marker on upgrade from versions < 8.8.0

## What does this PR do?



This PR checks if Agent is being run as part of an ongoing upgrade (i.e. being re-exec'd) and, if so, creates the installation marker file if the previous version of the Agent (the one being upgraded from) was one before the installation marker file was introduced. 

## Why is it important?



In v8.8.0, we introduced an installation marker file to indicate that an Agent was running as installed (this was necessary to implement the [custom installation `--base-path` feature](https://github.com/elastic/elastic-agent/pull/2500/)). When an installed Agent that's older than v8.8.0 is upgraded, this installation marker file is not present. In such cases, we need to create it manually post-upgrade.  Otherwise, the upgrade will be [unsuccessful](https://github.com/elastic/elastic-agent/issues/2645).

## Checklist



- [ ] ~I have added an integration test or an E2E test~ will be covered as part of https://github.com/elastic/elastic-agent/pull/2618

## How to test this PR locally

1. Install a version of Elastic Agent older than `8.8.0`.
   ```
   cd /tmp
   wget https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-8.2.3-darwin-aarch64.tar.gz
   tar xzf elastic-agent-8.2.3-darwin-aarch64.tar.gz
   cd elastic-agent-8.2.3-darwin-aarch64
   sudo ./elastic-agent install
   ```

2. Build Agent with this PR (should result in an Agent with version `8.9.0-SNAPSHOT`).
   ```
   cd /path/to/agent/repo
   DEV=true SNAPSHOT=true EXTERNAL=true PLATFORMS="darwin/arm64" PACKAGES="tar.gz" mage -v clean package
   ```

3. Upgrade to the Agent built with this PR.
   ```
   cd
   sudo elastic-agent upgrade --source-uri file:///path/to/agent/repo/build/distributions/ --skip-verify 8.9.0-SNAPSHOT
   ```

4. **Wait until the upgrade watcher has finished running.** This should be [10 minutes](https://github.com/elastic/elastic-agent/blob/29bf0ab601895fc498abf8724fc1918783956de5/internal/pkg/agent/cmd/watch.go#L33). Waiting ensures that the Agent is either successfully completed the upgrade or has been rolled back.  You can also keep checking the process list and wait until there is no `elastic-agent watch` process present.
   ```
   ps auxw | grep 'elastic-agent watch' | grep -v grep
   ```
   
5. Check that the new version of Agent is running.
   ```
   sudo elastic-agent version
   ```

6. Check that the running Agent is healthy.
   ```
   sudo elastic-agent status
   ```

## Related issues


- Fixes #2645

## Use cases



## Screenshots



## Logs



## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
